server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java

```diff
@@ -243,30 +266,33 @@ private void closeCurrentReader() {
         }
     }
 
-    private String getLine(BufferedReader reader, long startFrom, int retry) throws IOException {
+    private String getLine(BufferedReader reader) throws IOException {
         String line = null;
         try {
             while ((line = reader.readLine()) != null) {
                 currentLineNumber++;
-                if (currentLineNumber < startFrom) {
+                if (currentLineNumber < watermark) {
                     continue;
+                } else {
+                    watermark = 0;
                 }
                 if (line.length() == 0) {
                     continue;
                 }
                 break;
             }
-        } catch (SocketTimeoutException e) {
-            if (retry > MAX_SOCKET_TIMEOUT_RETRIES) {
-                URI uri = currentInput.uri();
-                LOGGER.error("Timeout during COPY FROM '" + uri.toString() + "' after " + retry + " retries", e);
-                throw e;
-            } else {
-                long startLine = retry == 0 ? currentLineNumber + 1 : startFrom;
+        } catch (SocketException | SocketTimeoutException e) {
+            if (backOffPolicy.hasNext()) {
+                watermark = watermark == 0 ? currentLineNumber + 1 : watermark;
                 closeCurrentReader();
                 initCurrentReader(currentInput, currentUri);
-                return getLine(currentReader, startLine, retry + 1);
+            } else {
+                URI uri = currentInput.uri();
+                LOGGER.error("Timeout during COPY FROM '" + uri.toString() +
+                             "' after " + MAX_SOCKET_TIMEOUT_RETRIES +
+                             " retries", e);
             }
+            throw e;
```

Comment:
Wouldn't it be cheaper while still valid to utilize a `NULL` return value as the indication for returning `FALSE` at `moveNext()`?  re-throwing and catching exceptions may be an expensive operation due to the stack trace frames...

Reply:
Just found out that the following change:
```
catch (IOException e) {
    if ((e instanceof SocketException || e instanceof SocketTimeoutException) && backOffPolicy.hasNext()) {
        return false;
    }
```
catches socket exceptions thrown while reading headers or [skipping lines](https://crate.io/docs/crate/reference/en/5.3/sql/statements/copy-from.html#skip) which makes this feature overlap with https://github.com/crate/crate/issues/13276. 
Unless we handle https://github.com/crate/crate/issues/13276 we should revert this change. What do you think? @seut